src/satosa/micro_services/ldap_attribute_store.py

```diff
@@ -112,189 +199,189 @@ def constructFilterValue(self, candidate, data):
 
-                responses = connection.response
-                satosa_logging(logger, logging.DEBUG, "{} LDAP server returned {} records".format(logprefix, len(responses)), context.state)
+                satosa_logging(logger, logging.DEBUG, "Querying LDAP server...", context.state)
+                message_id = connection.search(config['search_base'], search_filter, attributes=config['search_return_attributes'].keys())
+                responses = connection.get_response(message_id)[0]
+                satosa_logging(logger, logging.DEBUG, "Done querying LDAP server", context.state)
+                satosa_logging(logger, logging.DEBUG, "LDAP server returned {} records".format(len(responses)), context.state)
 
                 # for now consider only the first record found (if any)
                 if len(responses) > 0:
                     if len(responses) > 1:
-                        satosa_logging(logger, logging.WARN, "{} LDAP server returned {} records using IdP asserted attribute {}".format(logprefix, len(responses), identifier), context.state)
+                        satosa_logging(logger, logging.WARN, "LDAP server returned {} records using search filter value {}".format(len(responses), filter_val), context.state)
                     record = responses[0]
                     break
+        except LDAPException as err:
+            satosa_logging(logger, logging.ERROR, "Caught LDAP exception: {}".format(err), context.state)
+            return super().process(context, data)
+
+        except LdapAttributeStoreError as err:
+            satosa_logging(logger, logging.ERROR, "Caught LDAP Attribute Store exception: {}".format(err), context.state)
+            return super().process(context, data)
                         
         except Exception as err:
-            satosa_logging(logger, logging.ERROR, "{} Caught exception: {}".format(logprefix, err), context.state)
+            satosa_logging(logger, logging.ERROR, "Caught unhandled exception: {}".format(err), context.state)
             return super().process(context, data)
 
-        else:
-            satosa_logging(logger, logging.DEBUG, "{} Unbinding and closing connection to LDAP server".format(logprefix), context.state)
-            connection.unbind()
-
         # Before using a found record, if any, to populate attributes
         # clear any attributes incoming to this microservice if so configured.
-        if clear_input_attributes:
-            satosa_logging(logger, logging.DEBUG, "{} Clearing values for these input attributes: {}".format(logprefix, data.attributes), context.state)
+        if config['clear_input_attributes']:
+            satosa_logging(logger, logging.DEBUG, "Clearing values for these input attributes: {}".format(data.attributes), context.state)
             data.attributes = {}
 
         # Use a found record, if any, to populate attributes and input for NameID
         if record:
```

Comment:
This is a big `if-else` block. Maybe this logic should be split into smaller functions.